File: node_js/dos2unix.js
/*
dos2unix implementation with Node.js.
Copyright 2016 Sam Saint-Pettersen.

Released under the MIT License;
see LICENSE file.
*/

'use strict'

const fs = require('fs')

let USE_NATIVE = false

let native = null
try {
  native = require('./build/Release/dos2unix')
  USE_NATIVE = true
} catch (e) {
  USE_NATIVE = false
}

function isAscii (contents) {
  let ascii = true
  for (let i = 0; i < contents.length; i++) {
    if (contents.charCodeAt(i) > 127) {
      ascii = false
      break
    }
  }
  return ascii
}

function isDOS_eol (contents) {
  let dos_eol = false
  for (let i = 0; i < contents.length; i++) {
    if (contents.charAt(i) === '\r') {
      dos_eol = true
      break
    }
  }
  return dos_eol
}

function toUnixLineEndings (contents) {
  let ucontents = []
  for (let i = 0; i < contents.length; i++) {
    if (contents.charAt(i) !== '\r') {
      ucontents.push(contents.charAt(i))
    }
  }
  return ucontents.join('')
}

module.exports.dos2unix = function (filename, options) {
  if (options && options.native !== undefined) {
    USE_NATIVE = options.native
    if (native === null && USE_NATIVE) {
      USE_NATIVE = false
      console.warn(
      'dos2unix: Falling back to pure JS implementation ( native: ', USE_NATIVE, ')')
    }
  }

  let feedback = false
  if (options && options.feedback !== undefined) {
    feedback = options.feedback
  }

  let write = true
  if (options && options.write !== undefined) {
    write = options.write
    if (!write) {
      USE_NATIVE = false
    }
  }

  if (USE_NATIVE && native) {
    native.dos2unix(filename, feedback ? 1 : 0)
  } else {
    let message = 'dos2unix: File already has UNIX line endings or is binary.'
    let contents = fs.readFileSync(filename, 'utf8').toString()
    if (isAscii(contents) && isDOS_eol(contents)) {
      let converted = toUnixLineEndings(contents)
      if (write) {
        fs.writeFileSync(filename, converted, 'utf8', 'wb')
        return 0
      } else {
        if (feedback) {
          console.warn(message)
        }
        return converted
      }
    } else if (feedback) {
      console.warn(message)
    }
  }
  return 1
}

File: node_js/native/dos2unix.cc
/*
dos2unix implementation.
Copyright 2016 Sam Saint-Pettersen.

Native add-on.

Released under the MIT License;
see LICENSE file.
*/
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

bool is_ascii(string filename) {
    bool ascii = false;
    int c = -1;
    ifstream f(filename.c_str());
    while((c = f.get()) != EOF && c <= 127);
    if(c == EOF) {
        ascii = true;
    }
    f.close();
    return ascii;
}

bool is_dos_eol(string filename) {
    bool dos_eol = false;
    string l;
    ifstream f;
    f.open(filename.c_str(), ios::binary);
    while(getline(f, l)) {
        for(char& c : l) {
            if (c == '\r') {
                dos_eol = true;
                break;
            }
        }
    }
    f.close();
    return dos_eol;
}

string to_unix_line_endings(string contents) {
    string ucontents;
    for(char& c : contents) {
        if(c != '\r') ucontents.push_back(c);
    }
    return ucontents;
}

string dos2unix_str(string filename) {
    ifstream i;
    stringstream contents;
    i.open(filename, ios::binary);
    contents << i.rdbuf();
    i.close();
    if(is_ascii(filename) && is_dos_eol(filename)) {
        return to_unix_line_endings(contents.str());
    }
    return contents.str();
}

int dos2unix(string filename) {
    if(is_ascii(filename) && is_dos_eol(filename)) {
        ifstream i;
        stringstream contents;
        i.open(filename, ios::binary);
        contents << i.rdbuf();
        i.close();
        string out = to_unix_line_endings(contents.str());
        ofstream o;
        o.open(filename, ios::out | ios::binary);
        o << out;
        o.close();
    }
    else {
        // Trigger "File already has UNIX line endings or is binary" in calling code.
        return 1;
    }
    return 0;
}
